### ISSUE TITLE
[BUG] kubectl drain node is blocked by unexpected orphan engine processes

### ISSUE THREAD

**Issue by jsalatiel** [longhorn/longhorn#6552]
Just can not drain a node.  This happens almost at least half of the times I need to drain a node.
Cannot evict pod as it would violate the pod's disruption budget. 

## To Reproduce
Simple 4 node cluster, tried to drain one of the nodes.
I could see in the interface that all volumes had been detached and reattached and the pods started running in other nodes, but the node never finishes draining.
```kubectl get pods,pvc -o wide -A | grep "wp\|shar"
longhorn-system   pod/share-manager-pvc-64f91ac7-abcf-4ffa-8925-89de19818d3b            1/1     Running             0               16h     10.238.68.16     node2     <none>           <none>
longhorn-system   pod/share-manager-pvc-8325be46-d63e-4fc8-9fef-fa4f089d8490            1/1     Running             0               16h     10.238.73.14     node1     <none>           <none>
longhorn-system   pod/share-manager-pvc-93649f22-80e0-499d-bfeb-9aeb10c5ad55            1/1     Running             0               16h     10.238.67.11     node3     <none>           <none>
wp                pod/wp-mariadb-0                                                      1/1     Running             0               10m     10.238.67.15     node3     <none>           <none>
wp                pod/wp-wordpress-cfcc7fd6f-nh8fl                                      1/1     Running             0               10m     10.238.68.20     node2     <none>           <none>
wp                pod/wp-wordpress-cfcc7fd6f-nws8m                                      1/1     Running             3 (9m53s ago)   16h     10.238.73.16     node1     <none>           <none>
wp2               pod/wp-mariadb-0                                                      1/1     Running             0               16h     10.238.67.12     node3     <none>           <none>
wp2               pod/wp-wordpress-cfcc7fd6f-bdvzd                                      1/1     Running             1 (16h ago)     16h     10.238.73.17     node1     <none>           <none>
wp2               pod/wp-wordpress-cfcc7fd6f-sv8fw                                      1/1     Running             0               10m     10.238.68.19     node2     <none>           <none>
wp3               pod/wp-mariadb-0                                                      1/1     Running             0               10m     10.238.67.14     node3     <none>           <none>
wp3               pod/wp-wordpress-cfcc7fd6f-qw27r                                      1/1     Running             0               10m     10.238.68.18     node2     <none>           <none>
wp3               pod/wp-wordpress-cfcc7fd6f-wppst                                      1/1     Running             3 (9m51s ago)   16h     10.238.73.15     node1     <none>           <none>
wp          persistentvolumeclaim/data-wp-mariadb-0   Bound    pvc-efa3b148-a765-430f-a94e-a5df20757c0b   5Gi        RWO            longhorn       16h   Filesystem
wp          persistentvolumeclaim/wp-wordpress        Bound    pvc-93649f22-80e0-499d-bfeb-9aeb10c5ad55   2Gi        RWX            longhorn       16h   Filesystem
wp2         persistentvolumeclaim/data-wp-mariadb-0   Bound    pvc-d11ab7d8-6a5c-48b1-9260-4bb5b0043953   5Gi        RWO            longhorn       16h   Filesystem
wp2         persistentvolumeclaim/wp-wordpress        Bound    pvc-64f91ac7-abcf-4ffa-8925-89de19818d3b   2Gi        RWX            longhorn       16h   Filesystem
wp3         persistentvolumeclaim/data-wp-mariadb-0   Bound    pvc-51f9a96d-410f-49cd-be2d-d25ee33eac27   5Gi        RWO            longhorn       16h   Filesystem
wp3         persistentvolumeclaim/wp-wordpress        Bound    pvc-8325be46-d63e-4fc8-9fef-fa4f089d8490   2Gi        RWX            longhorn       16h   Filesystem

```

![image](https://github.com/longhorn/longhorn/assets/5522389/ad157b65-2296-482b-8d59-006ae6d0d49e)

![image](https://github.com/longhorn/longhorn/assets/5522389/b30d3ca0-f73e-4c76-8eb6-fdfece3493b3)

I used this command to drain:
```
kubectl drain --force --ignore-daemonsets --delete-emptydir-data --grace-period=10 node4
```

## Expected behavior

After all pods have been removed, the controller should remove the PDB that blocks draining of the manager allowing the drain to finish.

## Support bundle for troubleshooting

https://file.io/gfrW38dcRhJv

## Environment

 - Longhorn version: 1.5.1
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): kubespray
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: Onprem vanilla kubernetes installed by kubespray
   - Number of management node in the cluster: 3
   - Number of worker node in the cluster: 4

**Comment by ChanYiLin** [comment]
Thanks for the report
I can reproduce the issue in v1.5.1
The running replicas on the node should be stopped and shown as Failed. But nothing happened to my replicas.

@PhanLe1010 could you help to verify it? Thanks

**Comment by roger-ryao** [comment]
> Thanks for the report I can reproduce the issue in v1.5.1 The running replicas on the node should be stopped and shown as Failed. But nothing happened to my replicas.
> 
> @PhanLe1010 @roger-ryao could you help to verify it? Thanks

Hi @ChanYiLin & @PhanLe1010 
I attempted to reproduce the issue on my cluster, but I wasn't able to replicate it. I also tested the scenario described in https://github.com/longhorn/longhorn-tests/blob/05b21ac771aba74f3f0b2e68e0640678f5bcbc13/docs/content/manual/pre-release/upgrade/test-node-drain-policy.md under case 1.2.1 where there are multiple healthy replicas, and I didn't encounter the reported error. I plan to retest it again tomorrow.

**Comment by ChanYiLin** [comment]
Sorry for the misleading, last time I just mounted the volume to the host directly so I couldn't drain the node before because the volume could not be detached from the host automatically.

By mounting the volume to the pod, I can successfully drain the node because the pod will be evicted and rescheduled and the volume will be reattached to another node where pod recreated. Longhorn can then delete the pdb for the instance-manager to unblock the drain process.

**Comment by ChanYiLin** [comment]
I can see there is a engine `pvc-93649f22-80e0-499d-bfeb-9aeb10c5ad55-e-ce86048d` still running on node4, normally engine will only run on the same node where volume mounted.
But in your support bundle the engine ran on two instance-managers at the same time
```
  spec:
    image: longhornio/longhorn-instance-manager:v1.5.1
    nodeID: node4
    type: aio
  status:
    currentState: running
    instanceEngines:
      pvc-93649f22-80e0-499d-bfeb-9aeb10c5ad55-e-ce86048d:
        spec:
          backendStoreDriver: v1
          name: pvc-93649f22-80e0-499d-bfeb-9aeb10c5ad55-e-ce86048d
        status:
          state: running
          type: engine
```

```
  spec:
    image: longhornio/longhorn-instance-manager:v1.5.1
    nodeID: node3
    type: aio
  status:
    currentState: running
    instanceEngines:
      pvc-93649f22-80e0-499d-bfeb-9aeb10c5ad55-e-ce86048d:
        spec:
          backendStoreDriver: v1
          name: pvc-93649f22-80e0-499d-bfeb-9aeb10c5ad55-e-ce86048d
        status:
          state: running
          type: engine
```

**Comment by jsalatiel** [comment]
> I can see there is a engine `pvc-93649f22-80e0-499d-bfeb-9aeb10c5ad55-e-ce86048d` still running on node4, normally engine will only run on the same node where volume mounted.
> But in your support bundle the engine ran on two instance-managers at the same time
> ```
>   spec:
>     image: longhornio/longhorn-instance-manager:v1.5.1
>     nodeID: node4
>     type: aio
>   status:
>     currentState: running
>     instanceEngines:
>       pvc-93649f22-80e0-499d-bfeb-9aeb10c5ad55-e-ce86048d:
>         spec:
>           backendStoreDriver: v1
>           name: pvc-93649f22-80e0-499d-bfeb-9aeb10c5ad55-e-ce86048d
>         status:
>           state: running
>           type: engine
> ```
> 
> ```
>   spec:
>     image: longhornio/longhorn-instance-manager:v1.5.1
>     nodeID: node3
>     type: aio
>   status:
>     currentState: running
>     instanceEngines:
>       pvc-93649f22-80e0-499d-bfeb-9aeb10c5ad55-e-ce86048d:
>         spec:
>           backendStoreDriver: v1
>           name: pvc-93649f22-80e0-499d-bfeb-9aeb10c5ad55-e-ce86048d
>         status:
>           state: running
>           type: engine
> ```
> 

And what could have caused that?

**Comment by ChanYiLin** [comment]
I can't find the root cause from the log 
But base on your current situation, you can
1. uncorden node4
2. delete `instance-manager` pod on the node4 (`instance-manager-a88277b22f49670de0dada3c3706a2d1`)
3. wait until instance-manager pod is recreated, and check the instance-manager to see if there are still engine running in it.
```
kubectl get lhim/instance-manager-a88277b22f49670de0dada3c3706a2d1 -n longhorn-system -oyaml
```
4. if there is no engine running on it, you can try to drain the node again.

**Comment by docbobo** [comment]
I am having pretty much the same problem. Just sent you a support bundle from my setup - maybe that has some useful information.

**Comment by jsalatiel** [comment]
I keep having the same problem. The workaround of killing the node engine works, but it would be nice to have that fixed.

**Comment by docbobo** [comment]
One thing I've noticed: I have volumes showing as healthy (not Detached) even though I shutdown the pods using them today. And at least in one case, it's been the engines for those volumes that were kept running.

**Comment by ejweber** [comment]
I briefly checked @docbobo's support bundle. His cluster has a similar symptom to @jsalatiel's. There are two engines with the following behavior. Each:

- Is only represented by one Engine CR.
- Appears as running in two different InstanceManager CRs.

(`fd2f99b620d3-e` and `2a600e0d3b7a-e` are good search strings to find them if you are looking through the bundle or have access to the cluster).

**Comment by ejweber** [comment]
It seems pretty clear how an engine in the above scenario would prevent draining:

We check if we can delete the PodDisruptionBudget for the InstanceManager here:

https://github.com/longhorn/longhorn-manager/blob/8ce69c8f280a5a2715c528034e0274391ad761c0/controller/instance_manager_controller.go#L575-L582

Early on in that check, we try to determine whether all volumes are detached:

https://github.com/longhorn/longhorn-manager/blob/8ce69c8f280a5a2715c528034e0274391ad761c0/controller/instance_manager_controller.go#L697-L706

We do this by verifying that all instances are removed from the InstanceManager:

https://github.com/longhorn/longhorn-manager/blob/8ce69c8f280a5a2715c528034e0274391ad761c0/controller/instance_manager_controller.go#L826-L830

If there are any `instanceManager.status.instanceEngines`, we return false all the way up the chain:

https://github.com/longhorn/longhorn-manager/blob/8ce69c8f280a5a2715c528034e0274391ad761c0/controller/instance_manager_controller.go#L842-L846

The logic is sound, assuming there is not a "rogue" instance running within the InstanceManager. Unfortunately, we have two environments in which there is one.

**Comment by ejweber** [comment]
To anyone experiencing the kubectl drain issue this ticket was opened with, the following command is helpful in determining if you also have an extra engine process. In my test cluster, it shows I only have on engine running with name `pvc-d1d108a1-2cb9-42ca-b1e1-5eb8b3d49cbd-e-f8bb0399`, but all of the unique engine names any instance manager knows about with the number that are running.

```
-> kubectl get -n longhorn-system -o yaml lhim | grep 'name:.*-e-' | sort | uniq -c
      1           name: pvc-d1d108a1-2cb9-42ca-b1e1-5eb8b3d49cbd-e-f8bb0399
```

**Comment by docbobo** [comment]
Right now, the output shows `1` for every single volume. However, the list also contains volumes that definitely do not have a pod attached. 

I did a small experiment today and believe that if there would be an actual pod, the eviction of that pod would trigger the correct behavior. Just a guess though.

**Comment by ejweber** [comment]
Can you either send the names of those volumes to us via e-mail or post them here? There are some things I'd like to check, but I want to make sure I'm looking at the right volumes.

**Comment by ejweber** [comment]
> Right now, the output shows 1 for every single volume. However, the list also contains volumes that definitely do not have a pod attached.

Interesting, this bears further investigation.

I did manage to reproduce the inability to drain a node AFTER it is running a duplicate engine in its instance manager.

1. Install Longhorn v1.5.1.
2. Deploy a pod backed by a Longhorn volume (I actually created the volume through the UI and then deployed the workload later, but it should be the same).
3. Check instance manager logs for the engine process creation command/args.

```
-> kl logs --tail=-1 -l longhorn.io/component=instance-manager
[longhorn-instance-manager] time="2023-09-07T16:28:46Z" level=info msg="Creating instance" backendStoreDriver=v1 name=test-e-126ceb25 type=engine
[longhorn-instance-manager] time="2023-09-07T16:28:46Z" level=info msg="Creating process" args="[controller test --frontend tgt-blockdev --size 5368709120 --current-size 0 --engine-replica-timeout 8 --file-sync-http-client-timeout 30 --replica tcp://10.42.86.135:10000 --replica tcp://10.42.59.141:10000 --replica tcp://10.42.27.49:10000]" binary=/engine-binaries/longhornio-longhorn-engine-v1.5.1/longhorn name=test-e-126ceb25 portArgs="[--listen,0.0.0.0:]" portCount=1
[longhorn-instance-manager] time="2023-09-07T16:28:46Z" level=info msg="Process Manager: prepare to create process test-e-126ceb25"
[longhorn-instance-manager] time="2023-09-07T16:28:46Z" level=debug msg="Process Manager: validate process path: /engine-binaries/longhornio-longhorn-engine-v1.5.1/longhorn dir: /engine-binaries/ image: longhornio-longhorn-engine-v1.5.1 binary: longhorn"
[longhorn-instance-manager] time="2023-09-07T16:28:46Z" level=info msg="Process Manager: created process test-e-126ceb25"
```

4. Exec into a different instance manager and use the `longhorn-instance-manager` CLI to create a duplicate process. (This is certainly not how the duplicate processes are appearing, but its an easy recreate.)

```
instance-manager-0331339a5c09fb77fce790a98880c871:/ # longhorn-instance-manager process create --binary /host/var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.5.1/longhorn --name test-e-126ceb25 --port-args "--listen,0.0.0.0:" --port-count 1 controller test --frontend tgt-blockdev --size 5368709120 --current-size 0 --engine-replica-timeout 8 --file-sync-http-client-timeout 30

instance-manager-0331339a5c09fb77fce790a98880c871:/ # longhorn-instance-manager process ls
INFO[0000] Falling back to non tls client                error="failed to load tls key pair from file: open /tls-files/tls.crt: no such file or directory"
{
        "test-e-126ceb25": {
                "spec": {
                        "name": "test-e-126ceb25",
                        "binary": "/host/var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.5.1/longhorn",
                        "args": [
                                "controller",
                                "test",
                                "--frontend",
                                "tgt-blockdev",
                                "--size",
                                "5368709120",
                                "--current-size",
                                "0",
                                "--engine-replica-timeout",
                                "8",
                                "--file-sync-http-client-timeout",
                                "30",
                                "--listen",
                                "0.0.0.0:10010"
                        ],
                        "port_count": 1,
                        "port_args": [
                                "--listen,0.0.0.0:"
                        ]
                },
                "status": {
                        "state": "running",
                        "port_start": 10010,
                        "port_end": 10010
                }
        },
        ...
}
```

5. Confirm that there are now two instance managers running the same(ish) engine process. I didn't start the duplicate engine with access to the replicas, as I believe that would have failed.

```
-> kubectl get -n longhorn-system -o yaml lhim | grep 'name:.*-e-' | sort | uniq -c
      2           name: test-e-126ceb25
```


6. Try to `kubectl drain` the node with the "extra" engine process. This fails forever. Since there is a running engine process, the PDB can't be removed, but the engine object is tracking the original engine process, so longhorn-manager doesn't know to delete the "extra".
7. Try instead to `kubectl drain` the node with the original engine process. This works as expected.

**Comment by docbobo** [comment]
> Can you either send the names of those volumes to us via e-mail or post them here? There are some things I'd like to check, but I want to make sure I'm looking at the right volumes.

That would be

- pvc-2ff575f6-9ec9-48a9-9464-2ae546894060
- pvc-1efc9b3b-7621-47b1-a6e3-f8d99b5307f2

**Comment by docbobo** [comment]
Okay, I thought about this a little bit further. At least in my case, there seem to be some pending backup operations (see https://github.com/longhorn/longhorn/discussions/6281#discussioncomment-6434352). However, there aren't. 

So maybe the assumption about the absence if rogue instances isn't that wrong. Shall I file another ticket for the backups issue?

**Comment by schmidp** [comment]
I just now had a similar situation. `kured` was draining a node for a reboot. the pods were drained but one strictly-local volume stayed attached, even though the pod (cnpg) was stopped.

the moment I detached it via the UI (without force) the instance manager stopped and the node was able to finishing draining.

**Comment by ejweber** [comment]
Thanks @schmidp. Do you have a support bundle? For this issue, it is best to capture one while the drain is failing, as after the problem is resolved, we will not see the state of the `attachmentTickets` when detach was failing.

Note that if you have attached a volume via the UI, kubectl will not be able to drain it. Volumes attached via the UI must be detached via the UI.

**Comment by jsalatiel** [comment]
> Thanks @schmidp. Do you have a support bundle? For this issue, it is best to capture one while the drain is failing, as after the problem is resolved, we will not see the state of the `attachmentTickets` when detach was failing.
> 
> Note that if you have attached a volume via the UI, kubectl will not be able to drain it. Volumes attached via the UI must be detached via the UI.

On volumes attached via the UI, once detached via the UI, will it be auto reattached or will become a UI managed only volume forever? That does not look like a nice behavior. If i try to drain a node, i would expect it to be drained despite if it some volume was or not attached via the UI.

**Comment by schmidp** [comment]
> Thanks @schmidp. Do you have a support bundle? For this issue, it is best to capture one while the drain is failing, as after the problem is resolved, we will not see the state of the `attachmentTickets` when detach was failing.
> 
> Note that if you have attached a volume via the UI, kubectl will not be able to drain it. Volumes attached via the UI must be detached via the UI.

I have a support bundle now, would it be possible to send it via email? I feel a bit uneasy to share it here.

**Comment by innobead** [comment]
cc @c3y1huang

**Comment by c3y1huang** [comment]
@schmidp you can send it over to longhorn-support-bundle@Suse.com.

**Comment by ejweber** [comment]
Looking at the support bundle, I see two cordoned nodes. 

One of the nodes (...mea) has a couple of running engines, but they appear to be attached only by the CSI. It looks like they are just waiting for their corresponding workloads to be scaled down before they detach. For now, I am assuming this is not a concern.

The other node (...tvn) has a single running engine, `pvc-1a5253c5-f4e5-4729-b6fb-405bb33cee6c-e-57e230d7`. According to this volume's volumeattachment.longhorn.io, it has been attached via the UI.

```
spec:
  attachmentTickets:
    longhorn-ui:
      generation: 0
      id: longhorn-ui
      nodeID: ...tvn
      parameters:
        disableFrontend: "false"
        lastAttachedBy: "null"
      type: longhorn-api
  volume: pvc-1a5253c5-f4e5-4729-b6fb-405bb33cee6c
```

Assuming it was intentionally attached via UI, this is working as designed (though maybe not as expected). The fix is to detach it via the UI. @PhanLe1010, do you know if this is a behavioral change from v1.4.x (before we used attachment tickets)?

- If not, please feel free to open an enhancement request for this @schmidp. I'm not sure if anyone is relying on the current behavior or not, so we may need to introduce a setting (e.g. `DetachManuallyAttachedVolumesDuringDrain` or some better name). If we go this route, please provide as much justification as you can to help us prioritize (e.g. why are volumes manually attached in this environment?).
- If so, we probably need to treat this like a bug and figure out a mechanism for correcting it. In that case, we probably don't need the setting, as we are just reverting an undocumented behavioral change.

**Comment by schmidp** [comment]
@ejweber thank you for looking into this. sadly the pvc is now gone from our deployment, but what I can say for sure: there was no reason for attaching this manually except, that a few months ago we had a OOM situation where I afterwards tried to manually fix some pod issues by manually de- and attaching volumes.

I will wait for @PhanLe1010 reply and in case its intended behaviour request a setting as I believe its very easy to fall into this pit.

**Comment by PhanLe1010** [comment]
> Assuming it was intentionally attached via UI, this is working as designed (though maybe not as expected). The fix is to detach it via the UI. @PhanLe1010, do you know if this is a behavioral change from v1.4.x (before we used attachment tickets)?

Agree with @ejweber that the action to detach it via the UI. This behavior has been the same in older Longhorn versions: if user manually attach the volume using UI, the volume need to be detach from UI first before the workload on another node can attach the volume.

If user has never been using UI/Longhorn API to attach the volume, this is unexpected

**Comment by schmidp** [comment]
Assuming many of the mentioned problems are UI attached volumes, I wonder if it should be more discoverable if a volume has been attached via the UI. 

This could be solved by an icon or column in the UI, representing how it was attached.

**Comment by jsalatiel** [comment]
I like the icon idea, but i still think that drain should always have higher priority and detach even volumes attached by the UI.

**Comment by schmidp** [comment]
Personally, a setting (to not change existing behaviour) plus some kind of visualisation would also be my preference.

**Comment by ejweber** [comment]
I have opened https://github.com/longhorn/longhorn/issues/6978 to track the feature request for automatically detaching manually attached volumes during drain. 

Let's refocus this one on the original issue. Both @jsalatiel and @docbobo have clusters with multiple engine processes for a volume running on different nodes, and @docbobo has seen engine processes running for detached volumes. Another user has provided a support bundle with similar symptoms offline. We need to discover how Longhorn is "forgetting" to clean up old engine processes, as these processes prevent drains.

**Comment by user81230** [comment]
Haven't used attaching volumes via UI at all, but faced similar problems anyway. Most of the time, Longhorn doesn't evict instance-manager pods and thus blocking node draining. Maybe I should set bigger timeout (30m now) for draining nodes, but this is getting too time-consuming when maintaining bigger clusters. Only manually deleting such pods helps.

**Comment by ejweber** [comment]
I discovered a pretty straightforward reproduce of this behavior today related to [this code](https://github.com/longhorn/longhorn-manager/blob/8ce69c8f280a5a2715c528034e0274391ad761c0/controller/engine_controller.go#L531-L536) which still exists in `master`. Basically, if an instance-manager is in any state other than `running` (including `unknown`), longhorn-manager will not attempt to actually reach out to the instance-manager and stop an engine process before recording that it is stopped in its CR. But `unknown` is a state that instance-managers can come back from (e.g. if a kubelet is temporarily down, but then returns without the instance-manager pod actually stopping). When an instance-manager does, all of its processes remain running (and its CR remembers the ones that were never deleted).

Reproduce:

1. Deploy Longhorn `master-head`.
2. Create a volume (e.g. with the UI).
3. SSH to the node the volume is attached to.
4. Stop the kubelet.
5. Wait ~30s for Kubernetes to notice and for the instance-manager state to transition to `unknown`.
6. Detach the volume.
7. Start the kubelet.
8. Attach the volume to a different node.

Now the volume is in the state seen in the bundles of @docbobo and @jsalatiel. There is one engine CR tracking one engine process running on the NEW node, but the instance-manager CR on the OLD node also shows that same engine process running.

My personal preference is to never allow skipping the deletion of the engine process in an `unknown` state. Taking this approach would:

- Prevent us from accidentally orphaning processes in this case and ensure engine processes are definitely stopped when we think they are, but
- Cause the user to wait > 6min for Kubernetes to evict an instance-manager pod on an unrecoverable node for Kubernetes to evict the instance-manager pod and for its CR to transition to `error`. This is a state it cannot recover from without restarting, so we cannot have an orphan.

However, in a discussion with the North America team (@shuo-wu, @PhanLe1010, and @james-munson), the general direction we have decided to move in is to:

Create a PR 

- [ ] Create a PR to make a best-effort attempt to stop the engine process in this case to close out this issue. In the reproduce, there is actually no reason the longhorn-manager couldn't reach the instance-manager and make it happen.
- [ ] Follow up with the automated cleanup in this and other cases in https://github.com/longhorn/longhorn/issues/6764.
